pkg/nsx/services/inventory/inventory.go

```diff
@@ -0,0 +1,286 @@
+	inventoryService := &InventoryService{
+		requestBuffer:  make([]containerinventory.ContainerInventoryObject, 0),
+		pending_add:    make(map[string]interface{}),
+		pending_delete: make(map[string]interface{}),
+	}
+
+	// TODO, Inventory store should have its own store
+	inventoryService.applicationInstanceStore = &ApplicationInstanceStore{ResourceStore: commonservice.ResourceStore{
+		Indexer: cache.NewIndexer(keyFunc, cache.Indexers{string(ContainerApplicationInstance): indexFunc}),
+	}}
+	inventoryService.clusterStore = &ClusteStore{ResourceStore: commonservice.ResourceStore{
+		Indexer: cache.NewIndexer(keyFunc, cache.Indexers{string(ContainerCluster): indexFunc}),
+	}}
+	inventoryService.Service = service
+	return inventoryService, nil
+}
+
+func (s *InventoryService) Init() error {
+	_, err := s.GetContainerCluster()
+	// If there is no such cluster, create one.
+	// Otherwise, sync with NSX for different types of inventory objects.
+	if err != nil {
+		log.Error(err, "Cannot find existing container cluster, will create one")
+	} else {
+		return nil
+	}
+
+	newContainerCluster := s.BuildInentoryCluster()
+	_, err = s.AddContainerCluster(newContainerCluster)
+	if err != nil {
+		return err
+	}
+	log.Info("A new ContainerCluster is added", "cluster", newContainerCluster.DisplayName)
+	return nil
+}
+
+func (s *InventoryService) compareAndMergeUpdate(pre interface{}, cur interface{}) (string, map[string]interface{}) {
+	updateProperties := compareResources(pre, cur)
+	if pre == nil {
+		s.requestBuffer = append(s.requestBuffer, containerinventory.ContainerInventoryObject{ContainerObject: updateProperties, ObjectUpdateType: operationCreate})
+		return operationCreate, updateProperties
+	} else if len(updateProperties) > 2 {
+		s.requestBuffer = append(s.requestBuffer, containerinventory.ContainerInventoryObject{ContainerObject: updateProperties, ObjectUpdateType: operationUpdate})
+		return operationUpdate, updateProperties
+	} else {
+		return operationNone, nil
+	}
+}
+
+type empty struct{}
+type KeySet map[InventoryKey]empty
+
+func (s KeySet) Has(item InventoryKey) bool {
+	_, exists := s[item]
+	return exists
+}
+
+func (s KeySet) Insert(item InventoryKey) {
+	s[item] = empty{}
+}
+
+func (s KeySet) Delete(item InventoryKey) {
+	delete(s, item)
+}
+
+func (s *InventoryService) SyncInventoryObject(bufferedKeys KeySet) (KeySet, error) {
+	retryKeys := KeySet{}
+	startTime := time.Now()
+	defer func() {
+		log.Info("Finished syncing inventory object since ", "time", time.Since(startTime))
+	}()
+	for key := range bufferedKeys {
+		log.V(1).Info("Syncing inventory object", "object key", key)
+		externalId := key.ExternalId
+		namespace, name, err := cache.SplitMetaNamespaceKey(key.Key)
+		if err != nil {
+			log.Error(err, "Failed to split meta namespace key", "key", key)
+			continue
+		}
+		switch key.InventoryType {
+
+		case ContainerApplicationInstance:
```

Comment:
Better use a function or method to wrap these, we have much case code in the future.

Reply:
Done